### ISSUE TITLE
Function Tags

### ISSUE THREAD

**Issue by AliOsm** [emad-elsaid/rubyfunctions#24]
# Problem

- Users can't add tags to their functions, like: threading, string manipulation, regex, ...

# Suggested Solution

- Add a simple comma-separated tag field to allow users to add tags to their functions.

**Comment by emad-elsaid** [comment]
This one I want to make it automatically instead of the user do it manually,

we have the code, and we know what are the methods/modules were used in it, we can have the modules as tags, if he use `Queue` we add it as a tag, if he used `Array` we add it, and so on, so the code will be the king and the tags is the modules used as we do the work, user just need to add the code and got on with his life :smile:

**Comment by AliOsm** [comment]
It is a good idea to use the code as a source of tags, but what about tags that can't be extracted from the code. For example, function to calculate the shortest path between two nodes in a graph (BFS).

**Comment by emad-elsaid** [comment]
in your example what would be the tags? the user will add manually?

**Comment by AliOsm** [comment]
In my example, the user can add tags like BFS, graph or shortest path. We can't extract such tags from the code itself. Maybe we can extract them from the `usage` field, but this will need some machine learning to extract keywords from the text.

**Comment by emad-elsaid** [comment]
I see your point, so now lets assume adding tags is the user responsibility, what kind of constraints we'll have, should we limit tags to a specific set we seed to the website, allow users to add any kind of tags? maximum limit to tags per function? one tag? how to make sure tags are not duplicate, like "BFS", "bfs", "graph-bfs", "GraphBFS"? 
hummm, I lead towards a specific set of tags, but I'm not sure really

**Comment by AliOsm** [comment]
I think that we can go with user entered tags solution. while applying some constraints like preventing spaces and special characters, and processing the tags with applying text normalization like lower casing, steaming, non-pluralization, etc... Also, we can apply some tags merging techniques to merge semantically similar tags on weekly bases.

About the number of tags, I think it is ok to have a limit like 5 tags or something like that.

**Comment by emad-elsaid** [comment]
How about just allowing users to use Hashtags in their function usage, and in comments, in this case the'll be free to use it were and when they want, and will allow us to index functions by these hashtags, 
it even can be part of search feature, because all you need to do to get functions related to a tag is searching the DB for this hashtag 😃

**Comment by AliOsm** [comment]
Yes, the hashtags idea is great and it will allow users to be free in terms of where they put hashtags and how they put them. Any idea about the implementation?

**Comment by emad-elsaid** [comment]
there is a PR in the making now #42 after this is done we can start to add support for mentions and hashtags.

I think the markdown renderer can convert hashtags to links, links will lead to a search controller that returns functions that incluse this hashtag.
